src/MultiRegion/multi_region_split_explicit_free_surface.jl

```diff
@@ -53,20 +88,81 @@ function materialize_free_surface(free_surface::SplitExplicitFreeSurface, veloci
-    new_grid  = with_halo(new_halos, grid)
+    if extended_halos
+        old_halos = halo_size(getregion(grid, 1))
+        Nsubsteps = calculate_substeps(settings.substepping)
+
+        new_halos = multiregion_split_explicit_halos(old_halos, Nsubsteps+1, grid.partition)
+        new_grid  = with_halo(new_halos, grid)
+    else
+        new_grid = grid
+    end
 
     η = ZFaceField(new_grid, indices = (:, :, size(new_grid, 3)+1))
 
     return SplitExplicitFreeSurface(η,
                                     SplitExplicitState(new_grid, free_surface.settings.timestepper),
-                                    SplitExplicitAuxiliaryFields(new_grid),
+                                    SplitExplicitAuxiliaryFields(new_grid; extended_halos),
                                     free_surface.gravitational_acceleration,
                                     free_surface.settings)
 end
 
 @inline multiregion_split_explicit_halos(old_halos, step_halo, ::XPartition) = (max(step_halo, old_halos[1]), old_halos[2], old_halos[3])
 @inline multiregion_split_explicit_halos(old_halos, step_halo, ::YPartition) = (old_halos[1], max(step_halo, old_halo[2]), old_halos[3])
+
+const FillHaloSplitExplicit = SplitExplicitFreeSurface{<:Any, <:Any, <:SplitExplicitAuxiliaryFields{<:Any, <:Any, <:Nothing}}
+
+# Fallback!
+iterate_split_explicit!(free_surface, grid::MultiRegionGrids, Δτᴮ, weights, Nsubsteps) = 
+        @apply_regionally iterate_split_explicit!(free_surface, grid, Δτᴮ, weights, Nsubsteps)
+
+# Fill the halos after each substep
+iterate_split_explicit!(free_surface::FillHaloSplitExplicit, grid::MultiRegionGrids, Δτᴮ, weights, Nsubsteps) = 
+        fill_halo_iterate_split_explicit!(free_surface, grid, Δτᴮ, weights, Nsubsteps)
+
+function fill_halo_iterate_split_explicit!(free_surface, grid::MultiRegionGrids, Δτᴮ, weights, ::Val{Nsubsteps}) where Nsubsteps
+    arch = architecture(grid)
+
+    η         = free_surface.η
+    state     = free_surface.state
+    auxiliary = free_surface.auxiliary
+    settings  = free_surface.settings
+    g         = free_surface.gravitational_acceleration
+
+    # unpack state quantities, parameters and forcing terms 
+    U, V             = state.U,    state.V
+    Uᵐ⁻¹, Uᵐ⁻²       = state.Uᵐ⁻¹, state.Uᵐ⁻²
+    Vᵐ⁻¹, Vᵐ⁻²       = state.Vᵐ⁻¹, state.Vᵐ⁻²
+    ηᵐ, ηᵐ⁻¹, ηᵐ⁻²   = state.ηᵐ,   state.ηᵐ⁻¹, state.ηᵐ⁻²
+    η̅, U̅, V̅          = state.η̅, state.U̅, state.V̅
+    Gᵁ, Gⱽ, Hᶠᶜ, Hᶜᶠ = auxiliary.Gᵁ, auxiliary.Gⱽ, auxiliary.Hᶠᶜ, auxiliary.Hᶜᶠ
+
+    timestepper = settings.timestepper
+
+    η_args = (grid, Δτᴮ, η, ηᵐ, ηᵐ⁻¹, ηᵐ⁻², 
+              U, V, Uᵐ⁻¹, Uᵐ⁻², Vᵐ⁻¹, Vᵐ⁻², 
+              timestepper)
+
+    U_args = (grid, Δτᴮ, η, ηᵐ, ηᵐ⁻¹, ηᵐ⁻², 
+              U, Uᵐ⁻¹, Uᵐ⁻², V,  Vᵐ⁻¹, Vᵐ⁻²,
+              η̅, U̅, V̅, Gᵁ, Gⱽ, Hᶠᶜ, Hᶜᶠ, g, 
+              timestepper)
+
+    @unroll for substep in 1:Nsubsteps
+        averaging_weight = weights[substep]
+
+        fill_halo_regions!(U)
+        fill_halo_regions!(V)
```

Comment:
actually, this would work for multiregion, but I think it would not work for a cubed sphere where we need to fill `U` and `V` together

Reply:
@siddharthabishnu have you seen this? I think you need to correct this for your simulation